### ISSUE TITLE
cuBLAS support

### ISSUE THREAD

**Issue by dfdx** [JuliaGPU/GPUArrays.jl#34]
I'm interested to implement support for cuBLAS, but I haven't followed discussions lately so am a bit out of context and have a couple of questions. 

------

From the README:

> CUDAnative could support these easily as well, but we currently run into problems with the interactions of CUDAdrv and CUDArt

I see the relevant [commented coded](https://github.com/JuliaGPU/GPUArrays.jl/blob/d3d54c7a58154cd9727a2af96f01fd0c909a118e/src/backends/cudanative/cudanative.jl#L325-L344), but what exactly is wrong with it? 

------

Do we want cuBLAS to be integrated into CUDAnative.jl or just GPUArrays.jl?

**Comment by jekbradbury** [comment]
cuBLAS is already supported in the most recent branches of GPUArrays. And ideally none of the device libraries would be integrated into CUDAnative, I think.

**Comment by dfdx** [comment]
Cool! On what branch can I test it (current master still gives me an error)?

**Comment by jekbradbury** [comment]
Try https://github.com/JuliaGPU/GPUArrays.jl/tree/sd/for

**Comment by SimonDanisch** [comment]
you will need to checkout sd/for on Transpiler.jl and master on Sugar.jl
for OpenCL to work!

Am 28.06.2017 4:28 nachm. schrieb "jekbradbury" <notifications@github.com>:

**Comment by dfdx** [comment]
Thanks for the tips. I see that GPUArrays also depends on ModernGL which I'm having problems building. I'll report the results later.

**Comment by SimonDanisch** [comment]
ModernGL shouldn't have any binary dependencies! I should make that more clear in the readme, but if you have any errors while adding the package, there is a large probability that GPUArrays still works, since it just throws out backends that don't work, depending on the hardware.
It's just that the package manager doesn't support this kind of granularity, which is why it sadly prints errors while adding the package...

**Comment by dfdx** [comment]
I could have misinterpreted the output, because later I encountered issues with GLFW that are fixed now. However, with specified branches GPUArrays itself can't build anymore because of dependencies:

```
julia> using GPUArrays
INFO: Precompiling module GPUArrays.
WARNING: could not import Base.call into GLAbstraction

WARNING: deprecated syntax "abstract GPUArray{T,NDim}<:AbstractArray{T,NDim}" at /home/slipslop/.julia/v0.6/GLAbstraction/src/AbstractGPUArray.jl:21.
Use "abstract type GPUArray{T,NDim}<:AbstractArray{T,NDim} end" instead.

WARNING: deprecated syntax "abstract Unit" at /home/slipslop/.julia/v0.6/GLAbstraction/src/composition.jl:3.
Use "abstract type Unit end" instead.

WARNING: deprecated syntax "abstract Composable{unit}" at /home/slipslop/.julia/v0.6/GLAbstraction/src/composition.jl:4.
Use "abstract type Composable{unit} end" instead.
WARNING: both GeometryTypes and FixedSizeArrays export "Mat"; uses of it in module GLAbstraction must be qualified
ERROR: LoadError: LoadError: UndefVarError: Mat not defined
Stacktrace:
 [1] include_from_node1(::String) at ./loading.jl:569
 [2] include(::String) at ./sysimg.jl:14
 [3] include_from_node1(::String) at ./loading.jl:569
 [4] include(::String) at ./sysimg.jl:14
 [5] anonymous at ./<missing>:2
while loading /home/slipslop/.julia/v0.6/GLAbstraction/src/composition.jl, in expression starting on line 76
while loading /home/slipslop/.julia/v0.6/GLAbstraction/src/GLAbstraction.jl, in expression starting on line 57
ERROR: LoadError: Failed to precompile GLAbstraction to /home/slipslop/.julia/lib/v0.6/GLAbstraction.ji.
Stacktrace:
 [1] compilecache(::String) at ./loading.jl:703
 [2] _require(::Symbol) at ./loading.jl:456
 [3] require(::Symbol) at ./loading.jl:398
 [4] include_from_node1(::String) at ./loading.jl:569
 [5] include(::String) at ./sysimg.jl:14
 [6] anonymous at ./<missing>:2
while loading /home/slipslop/.julia/v0.6/Transpiler/src/Transpiler.jl, in expression starting on line 6
ERROR: LoadError: LoadError: LoadError: LoadError: Failed to precompile Transpiler to /home/slipslop/.julia/lib/v0.6/Transpiler.ji.
Stacktrace:
 [1] compilecache(::String) at ./loading.jl:703
 [2] _require(::Symbol) at ./loading.jl:456
 [3] require(::Symbol) at ./loading.jl:398
 [4] include_from_node1(::String) at ./loading.jl:569
 [5] include(::String) at ./sysimg.jl:14
 [6] include_from_node1(::String) at ./loading.jl:569
 [7] include(::String) at ./sysimg.jl:14
 [8] include_from_node1(::String) at ./loading.jl:569
 [9] include(::String) at ./sysimg.jl:14
 [10] include_from_node1(::String) at ./loading.jl:569
 [11] include(::String) at ./sysimg.jl:14
 [12] anonymous at ./<missing>:2
while loading /home/slipslop/.julia/v0.6/GPUArrays/src/backends/opencl/opencl.jl, in expression starting on line 17
while loading /home/slipslop/.julia/v0.6/GPUArrays/src/backends/supported_backends.jl, in expression starting on line 10
while loading /home/slipslop/.julia/v0.6/GPUArrays/src/backends/backends.jl, in expression starting on line 52
while loading /home/slipslop/.julia/v0.6/GPUArrays/src/GPUArrays.jl, in expression starting on line 13
ERROR: Failed to precompile GPUArrays to /home/slipslop/.julia/lib/v0.6/GPUArrays.ji.
Stacktrace:
 [1] compilecache(::String) at ./loading.jl:703
 [2] _require(::Symbol) at ./loading.jl:490
 [3] require(::Symbol) at ./loading.jl:398
```

------
Here's my `Pkg.status()` for mentioned packages.

```
 - CUBLAS                        0.1.0
 - CUDAdrv                       0.4.0
 - CUDAnative                    0.3.0+             master
 - CUDArt                        0.3.1
 - GPUArrays                     0.0.1+             sd/for
 - ModernGL                      0.1.1
 - CLBLAS                        1.1.1
 - CLFFT                         0.4.0
 - CUFFT                         0.0.4
 - GLAbstraction                 0.3.2
 - GLFW                          1.3.0+             master
 - OpenCL                        0.5.2
 - Sugar                         0.2.0+             master
 - Transpiler                    0.0.1+             sd/for
```
I'm not in a hurry, so if it's a known issue and you just need time to stabilize it, I don't mind waiting. On the other hand, if I can help somehow (testing or otherwise), just let me know.

**Comment by SimonDanisch** [comment]
Solved by #37
